### ISSUE TITLE
Centrality measures and graph clustering algorithms

### ISSUE THREAD

**Issue by helt** [jgrapht/jgrapht#367]
jgrapht does feature core graph theory algorithms, but (afaik) no graph analysis algorithms, like vertex/edge centrality measures (that are frequently used in social network analysis) or clustering methods like MCL and alike. 

If im right, then i would like to ask "why". Is it, because you (as authors) see this as out of scope? Or just because nobody cared?

**Comment by d-michail** [comment]
See the algorithms in the [org.jgrapht.alg.scoring](https://github.com/jgrapht/jgrapht/tree/master/jgrapht-core/src/main/java/org/jgrapht/alg/scoring) package. Some of them are pretty new. 

We are still missing vertex and edge centrality algorithms (aka an implementation of the algorithm of Brandes). If you have the time, consider contributing through a pull-request.

**Comment by jkinable** [comment]
In the past I've thought about adding some cluster algorithms, say for instance k-means clustering. 
Or, similar to that, there are a number of algorithms frequently used in the context of graph vizualization and projection which would have a similar structure.
For instance: given a weighted, not necessarily complete graph, where the weight on an edge (u,v) represent 'closeness', determine x-dimensional coordinates for each node such that the Euclidean distance between a pair of nodes reflects the closeness of the 2 nodes.

This process is a bit tricky though, since many of these algorithms can be implemented well with matrix multiplications. we would then need to include a math package with efficient matrix multiplication algorithms. Moreover, for efficiency, we would have to map the graph to a distance matrix. This is all nice, but the people who need this most likely don't use JGraphT as a data structure to maintain their graph. Instead, they would directly use a more compact, efficient representation such as a distance matrix. Consequently, I'm not so sure how useful it is to include this functionality in JGrapht.
Nevertheless, if there's a demand for this kind of algorithms, then this may be something we could consider.

**Comment by helt** [comment]
well, maybe, a submodule would provide a balance between core jgrapht algorithms and "Corner case" stuff.
Anyways, im not sure if im going beyond Label Propagation. I could just add it to my private project, but on the other hand, i would be happy to give something back. Just need to know the right place for it.

But i dont really agree on "compact representation" - graphs are per se more space-efficient than dense matrices. and convinient sparse matrix implementation for java are ... rare...

**Comment by jkinable** [comment]
Just to augment my previous answer, I certainly think there's a place for clustering/network analysis algorithms in Jgrapht. Here's a sketch of something that could be done:
1. create a new package 'clustering', i.e. jgrapht-core/.../alg/clustering
2. Design an interface for clustering algorithms and add it to alg/interfaces. You can probably duplicate the interface VertexColoringAlgorithm (in a sense, similar to clustering algorithms, Vertex Coloring is a partitioning of the nodes in the graph, so I presume the interface would be quite similar). Basic methods I would expect in the interface are:
- `int getNumberOfClusters()`
- `List<Set<V>> getClusters()`
- `Map<V, Integer> getClusterAssignment()`
-I'm not an expert on clustering algorithms so I'm not sure whether this interface would be adequate (it would be for simple clustering algorithms like k-means, dbscan, etc). Perhaps there exist clustering algorithms where a vertex can belong to multiple clusters, in which case the above interface wouldn't work well.
3. Implement clustering algorithms

Try to make the algorithms generic: i.e. works for both weighted and unweighted graphs. (Technically, all graphs in JgraphT are weighted, i.e. edge weights are always equal to 1.0 if no explicit weight has been set).

You may want to have a look at the `FloydWarshallShortestPaths` implementation. This implementation is structurally quite similar to some clustering algorithms. That is, the graph is mapped to a 2D distance matrix. The matrix is then used for calculations.

Finally, it is possible to use external packages (matrix/stats/etc) if necessary, but we try to use those as little as possible. 
Also, have a look at [Contributor guidelines](https://github.com/jgrapht/jgrapht/wiki/Contributor-Guidelines)

I'd recommend that you create a 'small' pull request first, containing a new cluster package, 1 interface, 1 clustering algorithm and corresponding test class. Once this is in place, it is usually easy to add several other clustering algorithms.

**Comment by weaver159** [comment]
Indeed. Clustering and partitioning algorithms would be interesting. The interface should be more general to allow a vertex to belong to different clusters. Something like incidence set:
`Pair<V,List<Integer>>`
should be returned, ie the list of cluster in which each vertex belongs. I might code the interface and simple clustering algorithm.
